File: cpp-api/sglocator.cpp
// C++11
#include <windows.h>
#include <iostream>
#include <string>
#include <vector>
#include <filesystem>
#include <fstream>
#include <regex>

using namespace std;

string getSteamLocation(BOOL& err, string& msg){
	err = false;
	HKEY hSubKey;
	DWORD dwType, dwSize = 255;
	char szBuffer[256];
	char* path = NULL;
	if (RegOpenKeyEx(HKEY_CURRENT_USER, "Software\\Valve\\Steam", 0, KEY_READ, &hSubKey) == ERROR_SUCCESS) {
		if (RegQueryValueEx(hSubKey, "SteamPath", NULL, &dwType, (LPBYTE)szBuffer, &dwSize) == ERROR_SUCCESS) {
			if (dwType == REG_SZ)
				path = szBuffer;
		}
		RegCloseKey(hSubKey);
	}
	if(path == NULL){
		err = true;
		msg = "Can not find steam installation path";
		return "";
	}else{
		return path;
	}
}

vector<string> getLibraryFolders(BOOL& err, string& msg){
	err = false;
	BOOL error;
	string error_msg;
	vector<string> libraryFolders;
	string steamLocation = getSteamLocation(error, error_msg);
	if(error){
		err = true;
		msg = error_msg;
		return libraryFolders;
	}
	string libraryFoldersVdf = steamLocation + "/config/libraryfolders.vdf";
	if(!filesystem::exists(libraryFoldersVdf)){
		err = true;
		msg = "File \"libraryfolders.vdf\" not found";
		return libraryFolders;
	}
	ifstream file(libraryFoldersVdf);
	if(!file){
		err = true;
		msg = "Open \"libraryfolders.vdf\" failed";
		return libraryFolders;
	}
	stringstream buffer;
	buffer << file.rdbuf();
	string content = buffer.str();
	
	// 正则匹配
	regex pattern("\"path\".+\"(.+)\"");
	std::smatch match;
	while (std::regex_search(content, match, pattern)) {
		string commonPath = match.str(1) + "\\steamapps\\common";
		if(filesystem::exists(commonPath) && filesystem::is_directory(commonPath)){
			libraryFolders.push_back(commonPath);
		}
		content = match.suffix();
	}
	return libraryFolders;
}

/**
 * @brief Get installation location of the game.
 * 
 * @param err: Setted to true when any error occurs.
 * @param msg: Message about error info.
 * @param gameFolderName: Must be the right folder name of the game, which you can get by right-click on the game in steam. 
 * 
 * @return locations: Paths where the game installed.
 * Cautions: It may contain more than one location because of the installation or uninstallation history.
 * Therefore, you need to verify for yourself which location is right.
 */
vector<string> getGameLocation(const string& gameFolderName, BOOL& err, string& msg){
	err = false;
	BOOL error;
	string error_msg;
	vector<string> locations;
	vector<string> folders = getLibraryFolders(error, error_msg);
	if(error){
		err = true;
		msg = error_msg;
		return locations;
	}
	for(auto it = folders.begin(); it != folders.end(); it++){
		string gamePath = *it + "\\" + gameFolderName;
		if(filesystem::exists(gamePath) && filesystem::is_directory(gamePath)){
			locations.push_back(gamePath);
		}
	}
	return locations;
}

File: js-api/sglocator.js
const winreg = require('winreg')
const fs = require('fs')

function getSteamLocation(){
    return new Promise((resolve, reject)=>{
        const regKey = new winreg({
            hive: winreg.HKCU,
            key: '\\Software\\Valve\\Steam'
        })
        regKey.get('SteamPath', (err, item)=>{
            if(err) throw new Error('Can Not Find Steam Installation Location')
            if(item){
                resolve(item.value)
            }else {
                throw new Error('Can Not Find Steam Installation Location')
            }
        })
    })
}

async function getLibraryFolders(){
    const steamLocation = await getSteamLocation()
    const libraryFoldersVdf = `${steamLocation}\\config\\libraryfolders.vdf`
    if(!fs.existsSync(libraryFoldersVdf)) throw new Error('libraryfolders.vdf Not Found')
    const content = fs.readFileSync(libraryFoldersVdf,{encoding: 'utf8'})
    const regex = /"path".+"(.+)"/g
    const result = content.matchAll(regex)
    const libraryFolders = []
    for (const item of result) {
        const commonPath = `${item[1]}\\steamapps\\common`
        if(!fs.existsSync(commonPath)) continue
        const stat = fs.statSync(commonPath)
        if(stat.isDirectory()){
            libraryFolders.push(commonPath)
        }
    }
    return libraryFolders
}

/**
 * Get installation location of the game.
 * @param gameFolderName Must be the right folder name of the game, which you can get by right-click on the game in steam.
 * @returns {Promise<*[]>} Paths where the game installed.
 *   Cautions: It may contain more than one location because of the installation or uninstallation history.
 *   Therefore, you need to verify for yourself which location is right.
 */
async function getGameLocation(gameFolderName){
    const folders = await getLibraryFolders()
    const locations = []
    for (const folder of folders) {
        const gamePath = `${folder}\\${gameFolderName}`
        if(!fs.existsSync(gamePath)) continue
        const stat = fs.statSync(gamePath)
        if(stat.isDirectory()){
            locations.push(gamePath)
        }
    }
    return locations
}

async function test(){

}
test()

module.exports = {
    getGameLocation
}


